File: python/tools/read_file.py
"""
Read file tool implementation.
Provides file reading functionality with schema caching.
"""

from typing import Any, Dict
from agent import ToolDefinition


class ReadFileTool(ToolDefinition):
    """
    Tool for reading file contents.
    Provides file reading functionality with cached schema for performance.
    """

    def __init__(self):
        # Cache the schema at initialization to avoid recomputation
        self._input_schema = {
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "The relative path of a file in the working directory."
                }
            },
            "required": ["path"]
        }

    def name(self) -> str:
        """Return the tool name."""
        return "read_file"

    def description(self) -> str:
        """Return the tool description."""
        return "Read the contents of a given relative file path. Use this when you want to see what's inside a file. Do not use this with directory names."

    def input_schema(self) -> Dict[str, Any]:
        """Return the cached JSON schema for tool input validation."""
        return self._input_schema

    def execute(self, path: str) -> str:
        """
        Execute the read file operation.

        Args:
            path: The relative path of the file to read

        Returns:
            The contents of the file as a string

        Raises:
            Exception: If the file cannot be read
        """
        try:
            with open(path, 'r', encoding='utf-8') as file:
                content = file.read()
            return content
        except Exception as e:
            raise Exception(f"Error reading file {path}: {str(e)}")

File: golang/tools/read_file.go
package tools

import (
	"encoding/json"
	"os"

	"github.com/anthropics/anthropic-sdk-go"
)

// ReadFileTool implements the read_file functionality
type ReadFileTool struct {
	inputSchema anthropic.ToolInputSchemaParam
}

// ReadFileInput represents the input parameters for the read_file tool
type ReadFileInput struct {
	Path string `json:"path" jsonschema_description:"The relative path of a file in the working directory."`
}

// NewReadFileTool creates a new instance of the read file tool
func NewReadFileTool() *ReadFileTool {
	return &ReadFileTool{
		inputSchema: GenerateSchema[ReadFileInput](),
	}
}

// Name returns the tool name
func (t *ReadFileTool) Name() string {
	return "read_file"
}

// Description returns the tool description
func (t *ReadFileTool) Description() string {
	return "Read the contents of a given relative file path. Use this when you want to see what's inside a file. Do not use this with directory names."
}

// InputSchema returns the JSON schema for tool input validation
func (t *ReadFileTool) InputSchema() anthropic.ToolInputSchemaParam {
	return t.inputSchema
}

// Execute reads the contents of a file at the specified path
func (t *ReadFileTool) Execute(input json.RawMessage) (string, error) {
	var readFileInput ReadFileInput
	err := json.Unmarshal(input, &readFileInput)
	if err != nil {
		return "", err
	}

	content, err := os.ReadFile(readFileInput.Path)
	if err != nil {
		return "", err
	}
	return string(content), nil
}
